<a href="https://colab.research.google.com/github/MGhanayim/qlora-finetuning-and-attention/blob/main/From_Finetuning_to_Attention_Inside_LLMs.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**Assignment Overview**

In this assignment, you will explore two fundamental aspects of modern NLP systems: fine-tuning large language models and understanding the attention mechanism that powers them.

**Learning Goals**

By the end of this assignment, you should be able to:

* Understand when and why to fine-tune language models
* Apply QLoRA for efficient adaptation of large models
* Analyze model outputs and limitations
* Explain and implement the attention mechanism
* Interpret attention patterns and their behavior

**Important Note**

* Do not modify or delete the task structure.
* Complete each task with:
   - Clean, well-organized code
   - Relevant visualizations
   - Clear insights and explanations
* Make sure to answer all required questions.
**Submission requirements:**
* Submit a **fully executed notebook** (all cells must run and outputs should be visible).
* There is no need to attach the training and test datasets as files, but you must present them as DataFrame tables within the notebook.


## Install Required Packages and Dataset

In [ ]:
# Task 1.2+ deps. peft is explicit (don't rely on trl pulling it in transitively).
# torchao deliberately omitted — SPEC uses NF4 via bitsandbytes, and torchao>=0.17 needs
# torch>=2.7, which cascades into transformers' import chain on older torch.
%pip install -q -U bitsandbytes trl peft accelerate datasets openai tqdm

In [2]:
import os
import sys
from getpass import getpass
from openai import OpenAI
from pydantic import BaseModel, ConfigDict, Field
import hashlib
from trl import SFTTrainer
import torch
import json as _json
from datasets import Dataset, load_dataset
import pandas as pd
import random
from transformers import (
    AutoTokenizer,
    AutoModelForCausalLM,
    AutoModelForSeq2SeqLM,
    BitsAndBytesConfig,
    TrainingArguments,
    Seq2SeqTrainingArguments,
    Seq2SeqTrainer,
    DataCollatorForSeq2Seq,
    pipeline,
    logging,
)
from peft import LoraConfig, TaskType, prepare_model_for_kbit_training, get_peft_model, PeftModel



import warnings
warnings.filterwarnings('ignore')


/opt/homebrew/Caskroom/miniconda/base/envs/nebius/lib/python3.11/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm


### Load API Keys, Secrets, Base URLs and Model Names

In [ ]:
def get_secret(key: str) -> str:
    """Fetch a secret from Colab userdata, local .env, or interactive prompt."""
    if "google.colab" in sys.modules:
        from google.colab import userdata
        value = userdata.get(key)
    else:
        from dotenv import load_dotenv
        load_dotenv()
        value = os.environ.get(key)

    if value is None:
        value = getpass(f"'{key}' not found in environment or secrets.\nEnter '{key}' manually: ")

    if not value:
        raise ValueError(f"'{key}' is empty.")
    return value


API_KEY_NAME = "NEBIUS_API_KEY"
api_key = get_secret(API_KEY_NAME)

NEBIUS_BASE_URL_NAME = "NEBIUS_BASE_URL"
nebius_base_url = get_secret(NEBIUS_BASE_URL_NAME)

Teacher_MODEL_NAME = "Qwen/Qwen3-30B-A3B-Instruct-2507"

### Colab bootstrap (optional — no-op locally)

If you run this notebook in Colab, the cell below clones this **private** repo so cached LLM outputs in `data/` (`pilot_raw.jsonl`, `test.jsonl`, eventually `train.jsonl`) are pulled in instead of regenerated. When running locally, the cell detects that and skips — the notebook is already inside the working tree.

**Colab URL**: [Open in Colab](https://colab.research.google.com/github/MGhanayim/qlora-finetuning-and-attention/blob/main/From_Finetuning_to_Attention_Inside_LLMs.ipynb)

**First-time Colab setup for private repos:**
1. *File → Open notebook → GitHub*, sign in, tick **"Include private repos"** so Colab can see the repo.
2. *Settings (⚙) → Secrets*, add `GH_TOKEN` — a GitHub personal access token with `repo` scope (and `NEBIUS_API_KEY`, `NEBIUS_BASE_URL` for the teacher LLM).
3. Run cells from the top — the bootstrap cell uses `get_secret("GH_TOKEN")` to authenticate the clone.


In [ ]:
# Colab bootstrap — clone the repo so cached LLM outputs in data/ (pilot_raw.jsonl,
# train.jsonl, test.jsonl) are reused instead of regenerated. No-op when running locally.
# Private repo: token comes from `get_secret("GH_TOKEN")` — set GH_TOKEN in Colab
# userdata or in `.env` locally.

if "google.colab" in sys.modules:
    import subprocess
    GH_USER = "MGhanayim"
    GH_REPO = "qlora-finetuning-and-attention"

    if not os.path.exists(GH_REPO):
        gh_token = get_secret("GH_TOKEN")
        # Build the full URL in Python (incl. token) and shell out via subprocess.
        # Avoids IPython's `!cmd {var}` substitution, which can silently fail to expand
        # Python variables inside conditional blocks on some IPython versions.
        clone_url = f"https://oauth2:{gh_token}@github.com/{GH_USER}/{GH_REPO}.git"
        subprocess.run(["git", "clone", clone_url], check=True)

    # os.chdir (not %cd) — works inside conditionals and matches the push cell below.
    os.chdir(GH_REPO)
    print(f"cwd: {os.getcwd()}")
    print("\ndata/ contents:")
    !ls -la data/
else:
    print("Running locally — skipping Colab bootstrap.")


## **Part 1 — QLoRA Fine-Tuning for Level-Adaptive Question Answering (75 points)**

In this part, you will fine-tune pretrained language models to answer questions at different explanation levels:

* Child — simple and intuitive explanation
* Student — clear educational explanation with moderate technical detail
* Expert — precise, technical, and domain-specific explanation

You will use a subset of the Databricks Dolly 15K dataset.

In [ ]:
dataset = load_dataset("databricks/databricks-dolly-15k", split="train[:5000]")


In [ ]:
def is_good_question(text):
    return any(q in text.lower() for q in [
        "why", "how", "what is", "explain"
    ])

filtered = [
    ex for ex in dataset
    if ex["category"] == "open_qa" and is_good_question(ex["instruction"])
]


In [ ]:
# Dedup by instruction text, then assign a stable per-question id.
# Dolly has a few copy-pasted entries among the keyword-filtered open_qa rows,
# so we drop duplicates before hashing — same instruction → same hash, and a
# repeated row would otherwise corrupt the resume-by-id logic downstream.
seen_instructions = set()
deduped = []
for ex in filtered:
    if ex["instruction"] not in seen_instructions:
        seen_instructions.add(ex["instruction"])
        deduped.append(ex)
dupes_dropped = len(filtered) - len(deduped)
filtered = deduped

# 8-char SHA256 of the instruction text — deterministic across re-runs.
for ex in filtered:
    ex["id"] = hashlib.sha256(ex["instruction"].encode()).hexdigest()[:8]

ids = [ex["id"] for ex in filtered]
assert len(set(ids)) == len(ids), f"Hash collisions after dedup: {len(ids) - len(set(ids))}"
print(f"{len(filtered)} unique questions ({dupes_dropped} duplicate instructions dropped). All IDs unique.")


577 unique questions (3 duplicate instructions dropped). All IDs unique.


In [ ]:
len(filtered), filtered[0]

### Dataset Samples by Quartiles

In [ ]:
filtered_df = pd.DataFrame(filtered).drop(columns=['context', 'category'])
filtered_df['response_len'] = filtered_df['response'].apply(len)
filtered_df['len_quartile'] = pd.qcut(filtered_df['response_len'], q=4, duplicates='drop')
sampled_df = filtered_df.groupby('len_quartile').sample(n=10, random_state=42).sort_values(by='len_quartile')

for _, row in sampled_df.iterrows():
    print(f"quartile:{row['len_quartile']}")
    print(f"Instruction: {row['instruction']}")
    print(f"Response length: {row['response_len']}")
    print(f"Response: {row['response']}")
    print("-" * 50)



quartile:(1.999, 104.0]
Instruction: What is React?
Response length: 91
Response: React is a JavaScript library that specializes in helping developers build user interfaces.
--------------------------------------------------
quartile:(1.999, 104.0]
Instruction: How many feet are in 1 mile?
Response length: 10
Response: 5,280 feet
--------------------------------------------------
quartile:(1.999, 104.0]
Instruction: What is Bart Simpson's best friend named?
Response length: 9
Response: Millhouse
--------------------------------------------------
quartile:(1.999, 104.0]
Instruction: What is the fastest way to travel between the United States and Croatia?
Response length: 86
Response: The fastest way to travel would be by airplane, which would be faster than a boat trip
--------------------------------------------------
quartile:(1.999, 104.0]
Instruction: What is the state capitol of Nevada?
Response length: 99
Response: The state capitol of Nevada is Carson City which was founded in 18

#### Quartile inspection — findings

We grouped the 577 filtered rows into 4 length-based quartiles and sampled 10 rows from each.

**Q1 (≤105 chars):**

Most Q1 rows (6 of 10 in the sample) are trivia questions, single fact lookups. Examples:
- "How many feet are in 1 mile?" → "5,280 feet"
- "What is Bart Simpson's best friend named?" → "Millhouse"

There's no concept to simplify or elaborate. The level-adaptation task is undefined for these questions.

**Q2 (105–264 chars):**

Q2 answers start to look more usable - they include definitions, comparisons. These can be worked into different levels 

**Q3 (264–476 chars):** 

Q3 has concept questions with answers that can be worked well into all three levels. 

**Q4 (>476 chars):** 

Q4 answers has some outliers with answers that are very long and detailed (e.g Debt Snowball and Avalanche at 1462 chars). 
A teacher LLM might unfaithfully rewrite that as a child answer, while omitting  / hallucinating.

#### Proposed further filtering
Min cap at 20 chars, max cap at 1000 chars. Filtering by length is imperfect, doesn't eliminate trivia question with long answers, but still it is very simple to implement.


In [ ]:
def is_good_answer_len(text):
    return 20 <= len(text) <=1000

length_filtered = [ex for ex in filtered if is_good_answer_len(ex["response"])]

len(length_filtered), length_filtered[0]


In [ ]:
filtered_df = pd.DataFrame(length_filtered).drop(columns=['context', 'category'])
filtered_df['response_len'] = filtered_df['response'].apply(len)
filtered_df['len_quartile'] = pd.qcut(filtered_df['response_len'], q=4, duplicates='drop')
sampled_df = filtered_df.groupby('len_quartile').sample(n=10, random_state=42).sort_values(by='len_quartile')

for _, row in sampled_df.iterrows():
    print(f"quartile:{row['len_quartile']}")
    print(f"Instruction: {row['instruction']}")
    print(f"Response length: {row['response_len']}")
    print(f"Response: {row['response']}")
    print("-" * 50)


### Test Nebius API for Schema-enforced output 

client = OpenAI(base_url=nebius_base_url, api_key=api_key)

#Pydantic model to validate the response
class AnimalColor(BaseModel):
    model_config = ConfigDict(extra="forbid")
    name: str
    color: str

print("AnimalColor.model_json_schema():",AnimalColor.model_json_schema())


def test_helper(test_name, response_format=None, pydantic_model=None):
    kwargs = {"response_format": response_format} if response_format is not None else {}
    print(f"--- {test_name} ---")
    
    try:
        if pydantic_model:
            response = client.beta.chat.completions.parse(
                model=Teacher_MODEL_NAME,
                temperature=0.0,
                max_tokens=200,
                messages=[{"role": "user", "content": "Invent a fictional animal. Return a JSON object with two fields: name (a made-up animal name) and color (its color). Do not include any other text."}],
                response_format=pydantic_model)
            validated = response.choices[0].message.parsed 
        else:
            response = client.chat.completions.create(
                model=Teacher_MODEL_NAME,
                temperature=0.0,
                max_tokens=200,
                messages=[{"role": "user", "content": "Invent a fictional animal. Return a JSON object with two fields: name (a made-up animal name) and color (its color). Do not include any other text."}],
                **kwargs)
            validated = AnimalColor.model_validate_json(response.choices[0].message.content)

        print("Validated response:", validated)
    except Exception as e:
        print("\tError:", str(e))
        return e.__class__.__name__
    print("\tSuccess")
    return True


test_helper("Test A - plain , no constraints", response_format=None)
test_helper("Test B -  json_object mode", response_format={"type": "json_object"})
test_helper("Test C - json_schema", response_format = {
    "type": "json_schema",
    "json_schema": {
        "name": "AnimalColor",
        "strict": True,
        "schema": AnimalColor.model_json_schema()
    }})
_ = test_helper("Test D - pydantic model", pydantic_model=AnimalColor)





### **Task 1.1 — Prepare a Custom Dataset (15 points)**

Create your own instruction-tuning dataset based on a subset of databricks-dolly-15k.

For each selected question-answer pair, create versions of the answer adapted to the requested expertise level.

Example format:

    Question: What is gradient descent?
    Expertise level: child
    Answer: Gradient descent is like walking downhill step by step until you reach the lowest point.

You should prepare:

1. Training set for fine-tuning -

   Use the filtered Dolly dataset as your source of questions.

   For each question, create examples in the example format per level.

2. Test set for evaluation -
    
   Create a small test set of 20 question-answer examples.

   The test set must:

  * include examples from all three expertise levels
  * be separate from the training set
  * be manually reviewed by you for quality
  * include answers that are appropriate for the requested expertise level

  Recommendation:

  1. Save the generated dataset as a '.jsonl' file so it can be loaded and reused later.

  2. Decide on the training format according to the model architecture:
    
    - For causal language models, such as `HuggingFaceTB/SmolLM2-360M-Instruct`, use one full text field:

    ```text
    ### Question:
    ...

    ### Expertise level:
    child / student / expert

    ### Answer:
    ...```

    - For seq2seq models, such as google/flan-t5-small, separate the input and target:

    input: Question + expertise level
    target: Answer
   
  3. Before generating the full dataset, test the pipeline on a small batch of examples to verify that:

  * the LLM returns valid JSON,
  * each question receives three expertise-level answers,
  * the saved .jsonl file can be loaded correctly,
  * the format matches the training code.

**Pydantic Schema and Prompt**

In [ ]:
# Pydantic model for leveled answers
class LeveledAnswers(BaseModel):
    """Three answers to the same question, calibrated to a child, a student, and an expert."""
    model_config = ConfigDict(extra="forbid")
    child: str = Field(description="Answer suitable for a child — simple and intuitive explanation, no jargon.")
    student: str = Field(description="Answer suitable for a student — clear educational explanation with moderate technical detail.")
    expert: str = Field(description="Answer suitable for an expert — precise, technical, and domain-specific explanation.")

print(LeveledAnswers.model_json_schema())


# Prompts for generating leveled answers.
# Server-side schema enforcement (via client.beta.chat.completions.parse) handles the
# JSON shape, so the prompts focus on register calibration and content rules, not format.

SYSTEM_PROMPT = """You rewrite a single question into three answers calibrated to three different readers: a child, a student, and an expert.

The three answers must cover the same factual content but be meaningfully distinct in vocabulary, depth, and length — not paraphrases of one another.

Level guidelines:
- child: a curious 8-year-old. 1–2 short sentences. Use a concrete analogy or everyday example. No jargon, no acronyms, no technical terms the child would not already know.
- student: a high-school or first-year undergraduate student. 2–4 sentences. Clear educational tone. Brief technical terms are fine; define them inline the first time they appear.
- expert: a practitioner in the field. 3–6 sentences. Use precise domain terminology without defining it. Be specific, accurate, and confident; avoid hedging and filler.

Rules for every answer:
- Answer the question directly. Do not restate the question.
- Do not refer to the audience inside the answer (no "for a child…", "as an expert…").
- No preambles ("Sure!", "Great question") and no sign-offs.
- Use the reference answer as a factual anchor. You may add accurate detail it omits; if it contradicts well-established knowledge, prefer the accurate knowledge.
"""

USER_PROMPT_TEMPLATE = """Question:
{question}

Reference answer (use for facts; do not copy its style or length):
{reference_answer}

Produce the three answers."""


{'additionalProperties': False, 'description': 'Three answers to the same question, calibrated to a child, a student, and an expert.', 'properties': {'child': {'description': 'Answer suitable for a child — simple and intuitive explanation, no jargon.', 'title': 'Child', 'type': 'string'}, 'student': {'description': 'Answer suitable for a student — clear educational explanation with moderate technical detail.', 'title': 'Student', 'type': 'string'}, 'expert': {'description': 'Answer suitable for an expert — precise, technical, and domain-specific explanation.', 'title': 'Expert', 'type': 'string'}}, 'required': ['child', 'student', 'expert'], 'title': 'LeveledAnswers', 'type': 'object'}


In [ ]:
# JSONL helpers — used by the pilot cell, the full-training-set generation cell, and
# any future generation pass. Kept here so the resume-by-id + append-on-success pattern
# has a single source of truth.

def load_jsonl_by_id(path: str) -> dict[str, dict]:
    """Load a JSONL file into a dict keyed by row['id']. Returns {} if the file is missing."""
    if not os.path.exists(path):
        return {}
    with open(path) as f:
        return {row["id"]: row for row in (_json.loads(line) for line in f if line.strip())}


def append_jsonl(path: str, row: dict) -> None:
    """Append one row as a JSONL line. Crash-safe basis for resume-by-id workflows."""
    with open(path, "a") as f:
        f.write(_json.dumps(row, ensure_ascii=False) + "\n")


**Test-set candidate sampling**

Pick 12 questions from the length-filtered pool, print each for manual review, and edit `kept_ids` / `rejected_ids` between cell runs to top up to 12 keepers. n=12 gives 36 candidates → 20 finals at ~55% per-level acceptance (vs. 67% at n=10 — tight if one register underperforms).

**Why manual cherry-pick, not just the random draw**

The test set is a benchmark, not a representative sample. With only 20 finals, one bad row = 5% of the eval signal, and a row with wrong ground truth or an ambiguous question makes base-vs-FT comparison meaningless (both models "fail" for reasons unrelated to fine-tuning).

The keyword + length filters caught the obvious junk, but a random draw still surfaces failure modes that defeat level-adaptation — seen in this batch:

- **procedural how-tos** — expert procedure = child procedure (e.g. vase arranging)
- **niche factoids** — child/student/expert collapse to the same one-line definition (e.g. paint vs pinto)
- **ambiguous questions** — no single correct answer (e.g. "largest airline" — by what metric?)
- **wrong ground truth** — anchor is incorrect (e.g. the "homonym" example is actually a homophone)
- **trivia / unit-conversion** — same number at every level (e.g. Gulf depth in Empire State buildings)
- **highly domain-specific** — child answer must define the domain first (e.g. windsurfing tack)

Each kept question must clear:

- one unambiguous correct answer
- three levels genuinely distinguishable in vocabulary and depth
- accurate ground truth

Rejected candidates stay in the training pool — training tolerates noise, the test set doesn't.


In [56]:
# Manual-review state. Edit these between cell runs as you review candidates.
kept_ids: set[str] = {
    "8302bce2", "05444020", "462609d0", "cd0f621b",          # round 1
    "3bb87092", "1fa68125", "904e99f5",                       # round 2
    "13a43ee6",                                               # round 3
                                                              # round 4: none
    "ecde93f2", "2a74685d",                                   # round 5
    "a026e67e",                                               # round 6
                                                              # round 7: none
                                                              # round 8: none
    "dee7df0b",                                               # round 9
}
rejected_ids: set[str] = {
    "998720af", "2aa1396c", "4ff7a0a1", "c71f1081",
    "b5dcb923", "c6f1e32d", "eb77d771", "166ba312",          # round 1
    "83ab3fcf", "f2b8f550", "8ef82e7a", "fb276768", "f9f50208",  # round 2
    "9b2395ce", "396054c4", "35a0c64f", "290d5014",          # round 3
    "e3fc7cc8", "02c4c0d9", "382260e6", "72f58ea3",          # round 4
    "250b24d0", "5ed7b954",                                   # round 5
    "8731c47e",                                                # round 6
    "b0e80bef",                                                   # round 7
    "80439c0f",                                                   # round 8
}

TARGET_N = 12

length_filtered_ids = [ex["id"] for ex in length_filtered]
by_id = {ex["id"]: ex for ex in length_filtered}

def draw_replacements():
    """Sample enough fresh candidates to bring `kept_ids + new` up to TARGET_N."""
    seen = kept_ids | rejected_ids
    eligible = [i for i in length_filtered_ids if i not in seen]
    need = TARGET_N - len(kept_ids)
    if need <= 0:
        return []
    # Seed is deterministic for reproducibility; shifts with the rejected count
    # so re-runs after new rejections don't keep proposing the same first candidates.
    rng = random.Random(42 + len(rejected_ids))
    return rng.sample(eligible, need)

candidate_ids = draw_replacements()

print(f"kept: {len(kept_ids)}/{TARGET_N}  |  rejected so far: {len(rejected_ids)}  |  new candidates this run: {len(candidate_ids)}")
print()

for i, qid in enumerate(candidate_ids, 1):
    ex = by_id[qid]
    print(f"--- [{i}] id={qid} ---")
    print(f"Q: {ex['instruction']}")
    print(f"A: {ex['response']}")
    print()

if len(kept_ids) == TARGET_N:
    # Sort for stable ordering — set iteration order isn't reproducible across runs.
    test_question_ids = sorted(kept_ids)
    print(f"Locked in {TARGET_N} test question IDs: {test_question_ids}")


kept: 12/12  |  rejected so far: 26  |  new candidates this run: 0

Locked in 12 test question IDs: ['05444020', '13a43ee6', '1fa68125', '2a74685d', '3bb87092', '462609d0', '8302bce2', '904e99f5', 'a026e67e', 'cd0f621b', 'dee7df0b', 'ecde93f2']


**Pilot batch — generate 3-level answers for the 12 test questions**

This doubles as the pilot for the prompts (first real-data run) and the source material for the test set: 12 questions × 3 levels = 36 candidate answers, which then get hand-curated down to the 20 final test examples.

- Teacher LLM: `Qwen/Qwen3-30B-A3B-Instruct-2507` on Nebius (validated via the AnimalColor test).
- Temperature 0.3 — mostly deterministic for prompt-tuning, with enough variation to avoid stilted phrasing at the child level.
- Sequential calls, in-memory storage in `pilot_results`. Threading + append-on-success JSONL come in for the full training run later.
- Watch the printed outputs for: register collapse (child reads like student), child answers that copy the reference verbatim, and audience-mentions inside the answer ("for a child…"). Those are the signals to revisit `SYSTEM_PROMPT`.


In [ ]:
PILOT_RAW_PATH = "data/pilot_raw.jsonl"


def generate_leveled_answers(question: str, reference_answer: str) -> LeveledAnswers:
    """Single teacher-LLM call → typed child/student/expert answers.
    `.parse()` enforces the schema server-side, so we don't need a retry-on-validation
    loop. At pilot scale we let transport errors crash the cell and re-run — the
    resume-by-id cache below means we don't lose work."""
    response = client.beta.chat.completions.parse(
        model=Teacher_MODEL_NAME,
        temperature=0.3,
        max_tokens=2000,  # defensive cap; expected ~250-600 tokens for full 3-level JSON
        messages=[
            {"role": "system", "content": SYSTEM_PROMPT},
            {"role": "user", "content": USER_PROMPT_TEMPLATE.format(
                question=question,
                reference_answer=reference_answer,
            )},
        ],
        response_format=LeveledAnswers,
    )
    return response.choices[0].message.parsed


# Resume-by-id: temp=0.3 outputs aren't seeded, so freezing the pilot to disk is what
# actually makes the test set reproducible across Colab/local runs.
os.makedirs("data", exist_ok=True)
cached_by_id = load_jsonl_by_id(PILOT_RAW_PATH)

pilot_results = [cached_by_id[qid] for qid in test_question_ids if qid in cached_by_id]
to_generate = [qid for qid in test_question_ids if qid not in cached_by_id]

print(f"Cached: {len(pilot_results)}  |  to generate: {len(to_generate)}")

for i, qid in enumerate(to_generate, 1):
    ex = by_id[qid]
    print(f"\n{'='*70}")
    print(f"[{i}/{len(to_generate)}] id={qid}")
    print(f"Q: {ex['instruction']}")
    ref_short = ex['response'][:200] + ('...' if len(ex['response']) > 200 else '')
    print(f"Reference: {ref_short}")

    answers = generate_leveled_answers(ex["instruction"], ex["response"])
    row = {
        "id": qid,
        "question": ex["instruction"],
        "reference_answer": ex["response"],
        "answers": answers.model_dump(),
    }
    pilot_results.append(row)
    append_jsonl(PILOT_RAW_PATH, row)

    print(f"\n  CHILD:   {answers.child}")
    print(f"  STUDENT: {answers.student}")
    print(f"  EXPERT:  {answers.expert}")

print(f"\n{'='*70}")
print(f"pilot_results: {len(pilot_results)} × 3 = {len(pilot_results) * 3} candidate answers")
print(f"Source of truth: {PILOT_RAW_PATH} (delete the file to regenerate everything)")


**Curate the test set — matched-pair design, 7 questions × 3 levels = 21 examples**

The pilot generated 36 candidate answers (12 questions × 3 levels). We curate down to 21 — the 7 questions that scored Strong/Excellent at *all three* levels in the pilot review.

**Why 21 and not the spec's 20:** strict per-level balance (7/7/7) plus matched-pair design — every question appears at all three levels, so per-level comparisons in Task 1.4 hold the question constant. An unmatched 20 (6/7/7 split with different questions per level) confounds level-adaptation with topic difficulty.

**Why these 7:** dropped the 3 questions whose child answer was weak (BJJ analogy felt preachy, advertising had no real analogy, ChatGPT's child was flat), plus 2 econ/finance questions (stock index, inflation control) to reduce the finance cluster. The 5 dropped questions go back to the training pool — test bar is stricter than training bar.


In [61]:
# Matched-pair test set: 7 questions × 3 levels = 21 examples.
# Slight deviation from the spec's "20" — justified by per-level balance (7/7/7) and
# matched-pair design (every question appears at all 3 levels, so per-level comparisons
# hold the question constant). The 7 below are the questions that scored Strong/Excellent
# at all three levels in the pilot review; the other 5 (BJJ, advertising, ChatGPT had
# weak child answers; stock index and inflation dropped for econ/finance balance) are
# released back to the training pool.
TEST_KEEP_IDS = [
    "cd0f621b",  # Prime number (math)
    "05444020",  # Von Neumann architecture (CS)
    "ecde93f2",  # SQL (CS/DB)
    "8302bce2",  # Terraform modules (DevOps)
    "1fa68125",  # Romance language (linguistics)
    "dee7df0b",  # Edgeworth box (econ)
    "462609d0",  # Treasury Bond (finance)
]
assert set(TEST_KEEP_IDS).issubset({r["id"] for r in pilot_results}), \
    "TEST_KEEP_IDS must be a subset of pilot_results ids"
assert len(set(TEST_KEEP_IDS)) == len(TEST_KEEP_IDS), "Duplicate id in TEST_KEEP_IDS"

# Build the 21-row flat test set (one row per (question, level, answer) tuple).
# Flat form matches the Task 1.3 eval table shape (Question | Level | ...).
by_pilot_id = {r["id"]: r for r in pilot_results}
test_rows = []
for qid in TEST_KEEP_IDS:
    r = by_pilot_id[qid]
    for level in ("child", "student", "expert"):
        test_rows.append({
            "id": qid,
            "question": r["question"],
            "level": level,
            "answer": r["answers"][level],
        })

assert len(test_rows) == 21, f"Expected 21 rows, got {len(test_rows)}"
per_level = {lvl: sum(1 for r in test_rows if r["level"] == lvl) for lvl in ("child", "student", "expert")}
assert per_level == {"child": 7, "student": 7, "expert": 7}, f"Unbalanced: {per_level}"

# Write data/test.jsonl
os.makedirs("data", exist_ok=True)
test_path = "data/test.jsonl"
with open(test_path, "w") as f:
    for row in test_rows:
        f.write(_json.dumps(row, ensure_ascii=False) + "\n")

print(f"Wrote {len(test_rows)} examples to {test_path} (per-level: {per_level})")

# Required by the assignment: present the test set as a DataFrame.
test_df = pd.DataFrame(test_rows)
test_df["answer"] = test_df["answer"].str.slice(0, 120) + "…"   # preview-only truncation
test_df


Wrote 21 examples to data/test.jsonl (per-level: {'child': 7, 'student': 7, 'expert': 7})


,id,question,level,answer
0,cd0f621b,What is a prime number?,child,Think of a prime number like a special kind of...
1,cd0f621b,What is a prime number?,student,A prime number is a positive integer greater t...
2,cd0f621b,What is a prime number?,expert,A prime number is an integer p > 1 that is div...
3,05444020,What is the Von Neumann architecture?,child,Think of a computer like a kitchen where the C...
4,05444020,What is the Von Neumann architecture?,student,The Von Neumann architecture is a design model...
5,05444020,What is the Von Neumann architecture?,expert,The Von Neumann architecture is a stored-progr...
6,ecde93f2,What is SQL?,child,SQL is like a special set of instructions you ...
7,ecde93f2,What is SQL?,student,"SQL, or Structured Query Language, is a domain..."
8,ecde93f2,What is SQL?,expert,SQL is a declarative programming language stan...
9,8302bce2,What is the purpose of using Terraform modules ?,child,Think of Terraform modules like LEGO blocks. Y...


**Colab pre-flight smoke test**

Pilot + test set are cached in `data/`, so on a clean Colab session none of the cells above actually exercise the Nebius API. This cell makes one fresh `.parse()` call on a known question to validate auth, base URL, and schema before the full ~490-row training run below.

Cost: ~$0.0002. No JSONL write — the cached pilot/test data is untouched.

In [ ]:
# Pre-flight: one fresh API call on a fixed question. Prints the 3 outputs so you can
# eyeball quality (and, on a fresh Colab session, confirm the API path works at all)
# before kicking off the full training-generation cell below.
SMOKE_QID = "cd0f621b"  # "What is a prime number?"
_ex = by_id[SMOKE_QID]
print(f"Q: {_ex['instruction']}")

_answers = generate_leveled_answers(_ex["instruction"], _ex["response"])
print(f"\n  CHILD:   {_answers.child}")
print(f"  STUDENT: {_answers.student}")
print(f"  EXPERT:  {_answers.expert}")
print("\nSmoke test passed — Nebius API + schema OK.")

**Full training-set generation**

Training pool = the length-filtered question pool minus the 7 test keepers (exact count printed below). The pilot questions that didn't make the test cut are reused as-is — same model, same prompt, same temperature, so no value in regenerating them.

- **Concurrency**: `ThreadPoolExecutor(max_workers=5)` with `as_completed`. Workers do the API calls; the main loop serializes JSONL writes, so no append races and no lock needed.
- **Persistence**: append-on-success to `data/train.jsonl`. Resume-by-id on re-run, so a mid-run crash loses at most the few requests currently in flight.
- **Failure policy**: log and skip. `.parse()` enforces the schema server-side, so a failure here is almost always a transport blip — re-running the cell retries any skipped ids.
- **Expected cost**: well under $0.20 at $0.30/M output tokens.

In [ ]:
import time
from concurrent.futures import ThreadPoolExecutor, as_completed

TRAIN_PATH = "data/train.jsonl"
MAX_WORKERS = 5

# Training pool: all length-filtered IDs minus the 7 test keepers.
# Rejected pilot candidates flow back into the training pool — test bar is stricter than training bar.
_test_keep_set = set(TEST_KEEP_IDS)
train_question_ids = [qid for qid in length_filtered_ids if qid not in _test_keep_set]
assert len(train_question_ids) + len(TEST_KEEP_IDS) == len(length_filtered_ids)

os.makedirs("data", exist_ok=True)

# First-time seeding: the 5 pilot questions not kept for the test set were already generated
# (same model/prompt/temperature). Copy them in rather than spending API calls to redo them.
if not os.path.exists(TRAIN_PATH):
    _train_id_set = set(train_question_ids)
    seeded = [r for r in pilot_results if r["id"] in _train_id_set]
    for r in seeded:
        append_jsonl(TRAIN_PATH, r)
    print(f"Seeded {TRAIN_PATH} with {len(seeded)} pilot rows.")

# Resume-by-id cache (we only need the id set here, not the rows).
done_ids = set(load_jsonl_by_id(TRAIN_PATH))

todo_ids = [qid for qid in train_question_ids if qid not in done_ids]
print(f"Pool: {len(train_question_ids)}  |  done: {len(done_ids)}  |  remaining: {len(todo_ids)}  |  workers: {MAX_WORKERS}")


def _generate_row(qid: str) -> dict:
    ex = by_id[qid]
    answers = generate_leveled_answers(ex["instruction"], ex["response"])
    return {
        "id": qid,
        "question": ex["instruction"],
        "reference_answer": ex["response"],
        "answers": answers.model_dump(),
    }


failures: list[tuple[str, str]] = []
start = time.time()

# `as_completed` yields in the main thread, so append_jsonl is called serially.
with ThreadPoolExecutor(max_workers=MAX_WORKERS) as pool:
    futures = {pool.submit(_generate_row, qid): qid for qid in todo_ids}
    for i, fut in enumerate(as_completed(futures), 1):
        qid = futures[fut]
        try:
            row = fut.result()
        except Exception as e:
            failures.append((qid, repr(e)))
            print(f"[{i}/{len(todo_ids)}] FAIL id={qid}: {e!r}")
            continue
        append_jsonl(TRAIN_PATH, row)
        if i % 25 == 0 or i == len(todo_ids):
            elapsed = time.time() - start
            rate = i / elapsed if elapsed else 0.0
            print(f"[{i}/{len(todo_ids)}] ok id={qid}  ({rate:.2f} q/s, {elapsed:.0f}s elapsed)")

elapsed = time.time() - start
print(f"\nDone in {elapsed:.0f}s. Successes: {len(todo_ids) - len(failures)}, failures: {len(failures)}")
if failures:
    print("Failed ids (re-run the cell to retry — they are not in the JSONL):")
    for qid, err in failures[:20]:
        print(f"  {qid}: {err}")

# Sanity: file reloads and every row has the 3 expected levels.
train_rows = list(load_jsonl_by_id(TRAIN_PATH).values())
levels_ok = all(set(r["answers"]) == {"child", "student", "expert"} for r in train_rows)
print(f"\n{TRAIN_PATH}: {len(train_rows)} rows  |  all 3 levels present: {levels_ok}")
print(f"Flat training examples (rows × 3 levels): {len(train_rows) * 3}")

# Required by the assignment: present the training set as a DataFrame. Sample 20 — the
# full ~1494-row table truncates to a 10-row head/tail by default; a random sample of
# 20 shows real variety across questions and the three levels.
train_flat = [
    {"id": r["id"], "question": r["question"], "level": lvl, "answer": r["answers"][lvl]}
    for r in train_rows for lvl in ("child", "student", "expert")
]
train_df = pd.DataFrame(train_flat)
train_df["answer"] = train_df["answer"].str.slice(0, 120) + "…"   # preview-only truncation
train_df.sample(20, random_state=42).sort_values(["id", "level"]).reset_index(drop=True)


**Push `data/` updates back to GitHub (Colab only — no-op locally)**

After the curation cell writes/refreshes `data/test.jsonl` (and later `data/train.jsonl`), this cell commits and pushes those changes from Colab back to the repo so they persist across sessions and are available the next time you `git pull` locally. Authenticates using the token already embedded in the clone's `.git/config` — nothing extra to set up.


In [ ]:
# Push data/ updates back to GitHub when running in Colab.
# Auth: the clone URL embedded the token in .git/config, so `git push` reuses it.
# No-op locally — just `git add data/ && git commit && git push` from a terminal.
if "google.colab" in sys.modules:
    GH_REPO = "qlora-finetuning-and-attention"

    # Defensive: ensure cwd is inside the repo before any git commands. The bootstrap
    # cell %cd's into it, but this also recovers from out-of-order cell runs / kernel
    # resets / a failed bootstrap.
    if not os.path.isdir(".git"):
        for path in (GH_REPO, f"/content/{GH_REPO}"):
            if os.path.isdir(os.path.join(path, ".git")):
                os.chdir(path)
                print(f"(cwd was outside repo; cd'd into {path})")
                break
        else:
            raise RuntimeError(
                f"Not inside a git repo and {GH_REPO}/.git not found. "
                "Run the Colab bootstrap cell at the top of the notebook first."
            )

    !git config user.email "mgnayim@gmail.com"
    !git config user.name "MGhanayim"
    !git add data/
    print("--- staged changes ---")
    !git status --short data/
    print("--- commit + push ---")
    !git commit -m "data: update from Colab run" || echo "(nothing to commit)"
    !git push
else:
    print("Running locally — push from Colab is a no-op (commit + push from a terminal).")


### **Task 1.2 — Fine-Tune Models with QLoRA (10 points)**

Fine-tune the models using QLoRA, meaning:

1. Load the base model in 4-bit quantization
2. Add LoRA adapters
3. Train only the adapter parameters

You should repeat the fine-tuning process for both models:

1. HuggingFaceTB/SmolLM2-360M-Instruct
2. google/flan-t5-small

Pay attention each model requires the dataset to be formated diffrently.

### **Task 1.3 — Evaluation (20 points)**

Evaluate both fine-tuned models on your 20-example test set.

For each model, compare:

* outputs before fine-tuning
* outputs after fine-tuning
* whether the answer matches the requested expertise level
* whether the answer is clear and relevant
* whether the answer stays faithful to the question

Evaluate using this table:

|Question|Level|Base Output|Fine-tuned Output|Expected Answer|Better after FT?|Notes|
|-------|-------|-------|-------|-------|-------|-------|


### **Task 1.4 — Model Comparison and Discussion (15 points)**

Compare the performance of the two models:

* Which model adapted better to the expertise levels?
* Which model produced clearer answers?
* Which model followed the requested format better?
* Did one model hallucinate more than the other?

Explain any differences you observe.

In your discussion, consider that:

* SmolLM2-360M-Instruct is a decoder-only instruction model
* flan-t5-small is an encoder-decoder instruction model
* Different architectures may behave differently on instruction-following and text generation tasks

### **Task 1.5 - Conceptual Questions (15 points)**

Answer the following questions:

1. What changed after fine-tuning?
   
   Discuss whether the model became better at adapting its answer to the requested expertise level.
2. Why is QLoRA more memory efficient?
   
   Explain the role of 4-bit quantization and LoRA adapters.
3. What happens if you increase the LoRA rank?
   
   Discuss the tradeoff between model capacity, memory usage, and overfitting risk.
4. Why use LoRA / QLoRA instead of prompt engineering?

      Discuss:

      * In what cases prompt engineering is sufficient
      * When fine-tuning becomes necessary
      * What advantages QLoRA provides over prompting
      * What are the trade-offs (cost, flexibility, control)

## **Part 2 — Understanding Attention (25 points)**

Goal:

Build intuition for how attention works.

Task:

You will implement a simple attention mechanism from scratch (PyTorch) and visualize its behavior.

### **Task 2.1 Implement Scaled Dot-Product Attention (5 points)**

Given:

* Query (Q)
* Key (K)
* Value (V)

Compute:

$$ Attention(Q,K,V) = softmax(\frac{QK^T}{\sqrt{d_k}})V$$

In [ ]:
import torch
import torch.nn.functional as F
import math

def scaled_dot_product_attention(Q, K, V, mask=None):
    """
    Implements Scaled Dot-Product Attention.

    Args:
        Q: (batch, heads, seq_len_q, d_k)
        K: (batch, heads, seq_len_k, d_k)
        V: (batch, heads, seq_len_v, d_v)
        mask: (batch, heads, seq_len_q, seq_len_k) or None

    Returns:
        output: (batch, heads, seq_len_q, d_v)
        attention_weights: (batch, heads, seq_len_q, seq_len_k)
    """


    #  Get dimension for scaling
    d_k = Q.size(-1)


    # Compute attention scores

    scores = # TODO: compute dot-product between Q and K^T


    # Scale the scores

    # TODO: divide scores by sqrt(d_k)


    #  Apply mask (if given)

    if mask is not None:
        # TODO: mask out invalid positions (set to -inf)
        pass


    # Softmax to get attention

    attention_weights = # TODO: apply softmax over last dimension


    # Compute weighted sum

    output = #TODO: multiply attention_weights with V

    return output, attention_weights

### **Task 2.2 Visualize Attention (5 points)**


Plot attention weights as a heatmap for the given sentence


In [ ]:
import torch
import matplotlib.pyplot as plt
import seaborn as sns
import math



sentence = "the cat sat on the mat"


tokens = # TODO: split sentence into tokens

seq_len = # TODO: compute sequence length



# Create dummy embeddings

d_model = 16


embeddings = # TODO: initialize random embeddings of shape (seq_len, d_model)

# Treat embeddings as Q, K, V
Q = embeddings
K = embeddings
V = embeddings



# Compute Scaled Dot-Product Attention

attention_weights = # TODO: compute attention weights (use function from above)



# Plot attention heatmap

plt.figure(figsize=(8, 6))

# TODO: plot heatmap using seaborn
# - use attention_weights
# - set xticklabels and yticklabels to tokens
# - choose a colormap (e.g., "viridis")

plt.title("Attention Weights Heatmap")

# TODO: label axes
# plt.xlabel(...)
# plt.ylabel(...)

# TODO: rotate ticks if needed

plt.show()



### **Task 2.3 Experiments (15 points)**

**Experiment 1 — Change One Word:**

1. Use a simple sentence, for example:
    the cat sat on the mat
2. Compute and plot the attention weights as a heatmap.
3. Change one word in the sentence, for example:
    the dog sat on the mat
4. Recompute and plot the attention heatmap.
5. Compare the two heatmaps and explain whether the attention pattern changed.

**Experiment 2 — Compare Different Attention Heads:**

Repeat the attention visualization using at least two different attention heads.

For each head, create separate projection matrices:

$$W_Q, W_K, W_V$$

Use them to compute:
$$Q=XW_Q, K=XW_K, V=XW_V$$

Then compute and plot the attention weights for each head.

**Questions**

Answer briefly:

1. Did changing one word affect the attention weights? Why or why not?
2. Do different attention heads focus on different tokens?
3. Why might multi-head attention be useful in Transformer models?